### ISSUE TITLE
VLC 3.0.7 doesn't build with s[r]libtool, clang and LTO

### ISSUE THREAD

**Issue by RJVB** [midipix-project/slibtool#18]
I tried to build VLC 3.0.7.1 as described in the subject, but this fails on libvlccore, both using the MAKEFLAGS=LIBTOOL= approach as when editing VLC's own libtool script to pass on all arguments to rlibtool (VLC uses wrapper scripts for libtool).

I think the failure hinges on 2 things that are probably interrelated:
- libcompat.[l]a is not generated using llvm-ar but using clang, thus probably creating a library that lacks an index (see the linker errors in the log).
- a shared libcompat library is created instead of a static one.

I'm attaching logs.

I tried restoring VLC's libtool script and linking libvlccore but I get the same error message. I need to recreate libcompat with the original libtool script in order for this to work.
However, I still couldn't link libvlccore using rlibtool with the correct libcompat; there's an unspecified file not found error (hint: always print the filename with such messages, given the amount of output (noise) that can surround the message).

Log using rlibtool:
[vlc-rlibtool.log](https://github.com/midipix-project/slibtool/files/3725364/vlc-rlibtool.log)

Log using the stock build system:
[vlc.log](https://github.com/midipix-project/slibtool/files/3725366/vlc.log)

**Comment by midipix** [comment]
Quickly looking at this, I see that several steps are performed by ../doltlibtool; perhaps that's what you meant when you mentioned 'the original libtool script', but in general, slibtool should _not_ be used in conjunction with other libtool wrappers. What I find conspicuous, for instance, is the following from vlc-rlibtool.log:

> ../doltlibtool  --tag=CC   --mode=link ccache /opt/local/bin/clang-mp-5.0  -O3 -g -flto=thin -m64  -Werror=unknown-warning-option -Werror=invalid-command-line-argument>
>
> rlibtool: link: ln -s /dev/null .libs/libcompat.a.disabled

Which suggests that the compile step is performed by ../doltlibtool, whereas the linking step is performed by slibtool

If possible, can you build from scratch using rdlibtool this time around? The 'd' stands for debugging, and will provide us with additional useful information (most notably the invocation arguments).

**Comment by RJVB** [comment]
A priori doltlibtool invokes the libtool script created by `configure` but you're right that I don't really know what it intends to do. I'll try building with `MAKEFLAGS=LIBTOOL=${prefix}/bin/libtool` to see if that wrapper is required.

Built with rdlibtool:
[vlc-rdlibtool.log](https://github.com/midipix-project/slibtool/files/3726309/vlc-rdlibtool.log)

**Comment by midipix** [comment]
`grep dolt vlc-rdlibtool.log` still suggests that the build uses a mix of gnu libtool and slibtool. Are `doltlibtool` and/or `doltcompile` somehow hard-coded into vlc's build system, or is it possible to override both from the command-line?

**Comment by RJVB** [comment]
>`grep dolt vlc-rdlibtool.log` still suggests that the build uses a mix of gnu libtool and slibtool. Are `doltlibtool` and/or `doltcompile` somehow hard-coded into vlc's build system, or is it possible to override both from the command-line?

I only see evidence of using doltcompile, which is indeed hardcoded. But that's not a wrapper for libtool. It seems to be a basic script to generate .lo files (and invokes the compiler directly).

**Comment by midipix** [comment]
doltcompile performs the task of libtool there, so ideally you should not be using it, but rather rdlibtool, for the compile step as well. But, let's ignore that for now and address one issue at a time, starting with libcompat.la.

As you might recall, r[d]libtool parses the libtool script generated via ./configure in order to determined whether static and/or dynamic builds are enabled. Can you run the following two commands in /opt/local/var/lnxports/build/_opt_local_linux-ports_multimedia_VLC/VLC/work/vlc-3.0.7.1/compat and post the output?

`./libtool --features`
`rlibtool --features`

If the libtool script is not present in that directory then use the one from ../ or ../../ -- more importantly, do _not_ use the stock /usr/bin/libtool as that one won't reflect project-specific preferences for intermediate/helper libraries.

**Comment by RJVB** [comment]
>doltcompile performs the task of libtool there, so ideally you should not be using it, but rather rdlibtool, for the compile step as well. But, let's ignore that for now

Yeah, in my experience it's best not to make this kind of suggestion to the VLC team...
I'm guessing they use that script to cut down on libtool overhead.

>As you might recall, r[d]libtool parses the libtool script generated via ./configure in order to determined whether static and/or dynamic builds are enabled.

Now that you mention it, and that must be the explanation. VLC builds at least 1 private library as a static, but also a number of public shared libraries (and a whole bunch of plugins/modules).

BTW, is that parsing affected when you add the following to the top of the project's libtool script, i.e. does rlibtool *parse* or *execute* the script?
```
exec /path/to/rlibtool "$@"
```

>Can you run the following two commands in /opt/local/var/lnxports/build/_opt_local_linux-ports_multimedia_VLC/VLC/work/vlc-3.0.7.1/compat and post the output?
>
>`./libtool --features`
```
host: x86_64-pc-linux-gnu
enable shared libraries
disable static libraries
```

>`rlibtool --features`
```



rdlibtool --features

host: x86_64-linux-gnu
enable shared libraries
disable static libraries
```

The toplevel libtool script indeed disables static libs, so the questions become:
- how come that it knows how to build libcompat as a static? I tried with `MAKEFLAGS=LIBTOOL=/path/to/vlc-3.0.7.1/libtool make libvlccore` and that works just as well as with doltlibtool. So that script isn't required, which is in line with my observation shunting libtool with the above exec trick isn't the solution (= doltlibtool doesn't add the missing magic).
- why doesn't the shared libcompat library provide the required symbols - visibility choices and lack of an export table?

In case you hadn't seen the doltlibtool code yet:
```
#!/bin/bash
top_builddir_slash="${0%%doltlibtool}"
: ${top_builddir_slash:=./}
args=()
modeok=false
tagok=false
for arg in "$@"; do
    case "$arg" in
        --mode=compile) modeok=true ;;
        --tag=CC|--tag=CXX) tagok=true ;;
        --tag=ASM|--tag=YASM) tagok=true; passthrough=true;;
        --silent|--quiet) ;;
        *) args[${#args[@]}]="$arg" ;;
    esac
done
if $modeok && $tagok ; then
    . ${top_builddir_slash}doltcompile "${args[@]}"
else
    exec ${top_builddir_slash}libtool "$@"
fi
```

Making it invoke rdlibtool instead of libtool on that exec line *also* doesn't provide a solution, so I think it's really safe to ignore the fact wrappers/alternatives are used (BUT we could presumably invoke r[d]libtool instead of doltcompile here.

**Comment by midipix** [comment]
> BTW, is that parsing affected when you add the following to the top of the project's libtool script, i.e. does rlibtool *parse* or *execute* the script?

The logic for that resides in slibtool's internal slbt_get_lconf_flags() interface; the starting point is the current working directory (and if not found then moving up the directory tree one level at a time), and so the result should not be affected by the physical location of rlibtool. Otherwise, your output of `rlibtool --features` wouldn't have matched that of the generated libtool script.

> The toplevel libtool script indeed disables static libs, so the questions become:
> - how come that it knows how to build libcompat as a static? I tried with `MAKEFLAGS=LIBTOOL=/path/to/vlc-3.0.7.1/libtool make libvlccore` and that works just as well as with doltlibtool. So that script isn't required, which is in line with my observation shunting libtool with the above exec trick isn't the solution (= doltlibtool doesn't add the missing magic).

Was about to ask the same... can you create a tarball of your entire build directory tree and upload it somewhere so that I could take a look?

**Comment by RJVB** [comment]
On Tuesday October 15 2019 05:17:17 midipix wrote:
>> BTW, is that parsing affected when you add the following to the top of the project's libtool script, i.e. does rlibtool *parse* or *execute* the script?
>
>The logic for that resides in slibtool's internal slbt_get_lconf_flags() interface; the starting point is the current working directory (and if not found then moving up the directory tree one level at a time), and so the result should not be affected by the physical location of rlibtool.

Not what I meant with "parse" vs. "execute". I.e. does rlibtool look at the libtool *source* or at the libtool *output*?

>Was about to ask the same... can you create a tarball of your entire build directory tree and upload it somewhere so that I could take a look?

Will do (after I walk the dog :)) but in the meantime I finally stopped overlooking the obvious:

```
+ func_infer_tag ccache /opt/local/bin/clang-mp-5.0 -O3 -g -flto=thin -m64 -Werror=unknown-warning-option -Werror=invalid-command-line-argument -I/opt/local/var/lnxports/build/_opt_local_linux-ports_multimedia_VLC/VLC/work/vlc-3.0.7.1/contrib/include -Wall -Wextra -Wsign-compare -Wundef -Wpointer-arith -Wvolatile-register-var -Wformat -Wformat-security -Wbad-function-cast -Wwrite-strings -Wmissing-prototypes -Werror-implicit-function-declaration -Winit-self -pipe -fvisibility=hidden -O3 -fno-math-errno -funsafe-math-optimizations -funroll-loops -fomit-frame-pointer -fstack-protector-strong -no-undefined \
     -static \
-L/opt/local/lib -Wl,--enable-new-dtags -Wl,-rpath,/opt/local/lib -O3 -g -flto=thin -Wl,-rpath=/opt/local/lib -Wl,-rpath,/opt/local/libexec/qt5/lib -L/opt/local/var/lnxports/build/_opt_local_linux-ports_multimedia_VLC/VLC/work/vlc-3.0.7.1/contrib/lib -Wl,-z,defs -o libcompat.la -rpath /opt/local/lib/vlc dummy.lo strlcpy.lo strnstr.lo -lm
```

If r[d]libtool somehow decides to ignore a `-static` link argument then that must be a bug, no?

**Comment by midipix** [comment]
> Not what I meant with "parse" vs. "execute". I.e. does rlibtool look at the libtool *source* or at the libtool *output*?

Pardon; rlibtool looks at the source; no execution of the actual libtool script takes place in that (or any other) context.

> If r[d]libtool somehow decides to ignore a `-static` link argument then that must be a bug, no?

Arguably yes; will take a closer look at that alongside the rest of the build tree once you have it:=)

**Comment by RJVB** [comment]
>Arguably yes; will take a closer look at that alongside the rest of the build tree once you have it:=)

https://we.tl/t-7ItYVmog9o

**Comment by midipix** [comment]
> -rpath /opt/local/lib/vlc 

This is the problem, and that seems to be a bug in vlc. For the relevant text in gnu libtool's documentation, search for `create a libtool static convenience library`. Documentation can be found here:

https://www.gnu.org/software/libtool/manual/libtool.html

Once you remove the offending -rpath argument you will get static "convenience" libraries built by either rlibtool or gnu libtool.

**Comment by RJVB** [comment]
On Wednesday October 16 2019 04:53:36 midipix wrote:
>> -rpath /opt/local/lib/vlc 
>
>This is the problem, and that seems to be a bug in vlc. 

Actually no, because

>For the relevant text in gnu libtool's documentation, search for `create a libtool static convenience library`. Documentation can be found here:
>>https://www.gnu.org/software/libtool/manual/libtool.html

That text says `you can omit the -rpath flag` - *can*, not *have to*. Which is why libtool itself has no problems here; it probably just ignores `-rpath` if `-static` is present.

**Comment by midipix** [comment]
Looks like our messages crossed... I see your point, and will post a patch later. I'll double-check that it doesn't break any of the packages included with midipix_build before I push it, though.

**Comment by midipix** [comment]
As of commit e86d8c08, explicit -static and/or --disable-shared are now "stronger" than -rpath. This should fix the above libcompat.la generation issue.

**Comment by RJVB** [comment]
>As of commit e86d8c08, explicit -static and/or --disable-shared are now "stronger" than -rpath. This should fix the above libcompat.la generation issue.

Sadly, no - hopefully almost but problems remain.

Trying as described earlier by doing `exec rdlibtool $@` at the top of the project libtool script and running inside the compat directory:

```
> ../doltlibtool  --tag=CC   --mode=link ccache /opt/local/bin/clang-mp-5.0  -O3 -g -flto=thin -m64  -Werror=unknown-warning-option -Werror=invalid-command-line-argument -I/opt/local/var/lnxports/build/_opt_local_linux-ports_multimedia_VLC/VLC/work/vlc-3.0.7.1/contrib/include -Wall -Wextra -Wsign-compare -Wundef -Wpointer-arith -Wvolatile-register-var -Wformat -Wformat-security -Wbad-function-cast -Wwrite-strings -Wmissing-prototypes -Werror-implicit-function-declaration -Winit-self -pipe -fvisibility=hidden -O3 -fno-math-errno -funsafe-math-optimizations -funroll-loops -fomit-frame-pointer -fstack-protector-strong -no-undefined -static -L/opt/local/lib -Wl,--enable-new-dtags -Wl,-rpath,/opt/local/lib -O3 -g -flto=thin -Wl,-rpath=/opt/local/lib -Wl,-rpath,/opt/local/libexec/qt5/lib -L/opt/local/var/lnxports/build/_opt_local_linux-ports_multimedia_VLC/VLC/work/vlc-3.0.7.1/contrib/lib -Wl,-z,defs -o libcompat.la -rpath /opt/local/lib/vlc dummy.lo strlcpy.lo strnstr.lo  -lm



/opt/local/bin/rdlibtool --tag=CC --mode=link ccache /opt/local/bin/clang-mp-5.0 -O3 -g -flto=thin -m64 -Werror=unknown-warning-option -Werror=invalid-command-line-argument -I/opt/local/var/lnxports/build/_opt_local_linux-ports_multimedia_VLC/VLC/work/vlc-3.0.7.1/contrib/include -Wall -Wextra -Wsign-compare -Wundef -Wpointer-arith -Wvolatile-register-var -Wformat -Wformat-security -Wbad-function-cast -Wwrite-strings -Wmissing-prototypes -Werror-implicit-function-declaration -Winit-self -pipe -fvisibility=hidden -O3 -fno-math-errno -funsafe-math-optimizations -funroll-loops -fomit-frame-pointer -fstack-protector-strong -no-undefined -static -L/opt/local/lib -Wl,--enable-new-dtags -Wl,-rpath,/opt/local/lib -O3 -g -flto=thin -Wl,-rpath=/opt/local/lib -Wl,-rpath,/opt/local/libexec/qt5/lib -L/opt/local/var/lnxports/build/_opt_local_linux-ports_multimedia_VLC/VLC/work/vlc-3.0.7.1/contrib/lib -Wl,-z,defs -o libcompat.la -rpath /opt/local/lib/vlc dummy.lo strlcpy.lo strnstr.lo -lm

rdlibtool: link: ln -s /dev/null .libs/libcompat.a.disabled
rdlibtool: link: ln -s /dev/null .libs/libcompat.so.disabled
rdlibtool: link: ln -s ../libcompat.la .libs/libcompat.la
rdlibtool: link: ln -s ../libcompat.la .libs/libcompat.lai
```

I'm not certain what command is executed by rdlibtool here, but the end result is both .a.disabled and a .so.disabled file that are symlinks to /dev/null.

Removing the -rpath argument I notice another issue:

```
> ../doltlibtool --tag=CC --mode=link ccache /opt/local/bin/clang-mp-5.0 -O3 -g -flto=thin -m64 -Werror=unknown-warning-option -Werror=invalid-command-line-argument -I/opt/local/var/lnxports/build/_opt_local_linux-ports_multimedia_VLC/VLC/work/vlc-3.0.7.1/contrib/include -Wall -Wextra -Wsign-compare -Wundef -Wpointer-arith -Wvolatile-register-var -Wformat -Wformat-security -Wbad-function-cast -Wwrite-strings -Wmissing-prototypes -Werror-implicit-function-declaration -Winit-self -pipe -fvisibility=hidden -O3 -fno-math-errno -funsafe-math-optimizations -funroll-loops -fomit-frame-pointer -fstack-protector-strong -no-undefined -static -L/opt/local/lib -Wl,--enable-new-dtags -Wl,-rpath,/opt/local/lib -O3 -g -flto=thin -Wl,-rpath=/opt/local/lib -Wl,-rpath,/opt/local/libexec/qt5/lib -L/opt/local/var/lnxports/build/_opt_local_linux-ports_multimedia_VLC/VLC/work/vlc-3.0.7.1/contrib/lib -Wl,-z,defs -o libcompat.la dummy.lo strlcpy.lo strnstr.lo -lm                              



/opt/local/bin/rdlibtool --tag=CC --mode=link ccache /opt/local/bin/clang-mp-5.0 -O3 -g -flto=thin -m64 -Werror=unknown-warning-option -Werror=invalid-command-line-argument -I/opt/local/var/lnxports/build/_opt_local_linux-ports_multimedia_VLC/VLC/work/vlc-3.0.7.1/contrib/include -Wall -Wextra -Wsign-compare -Wundef -Wpointer-arith -Wvolatile-register-var -Wformat -Wformat-security -Wbad-function-cast -Wwrite-strings -Wmissing-prototypes -Werror-implicit-function-declaration -Winit-self -pipe -fvisibility=hidden -O3 -fno-math-errno -funsafe-math-optimizations -funroll-loops -fomit-frame-pointer -fstack-protector-strong -no-undefined -static -L/opt/local/lib -Wl,--enable-new-dtags -Wl,-rpath,/opt/local/lib -O3 -g -flto=thin -Wl,-rpath=/opt/local/lib -Wl,-rpath,/opt/local/libexec/qt5/lib -L/opt/local/var/lnxports/build/_opt_local_linux-ports_multimedia_VLC/VLC/work/vlc-3.0.7.1/contrib/lib -Wl,-z,defs -o libcompat.la dummy.lo strlcpy.lo strnstr.lo -lm

rdlibtool: link: /opt/local/bin/clang-mp-ar crs .libs/libcompat.a .libs/dummy.o .libs/strlcpy.o .libs/strnstr.o
rdlibtool: exec error upon slbt_exec_link_create_archive(), line 1244: (see child process error messages).
rdlibtool: < returned to > slbt_exec_link(), line 1806.
Exit 2
```

Here rlibtool errors in deducing the name of the AR command. The llvm archiver is called llvm-ar (would be llvm-ar-mp-5.0 here) but the correct setting can (and should) be read from the AR= definition in the libtool script.
The good news is that I get a valid libcompat.a if I symlink clang-mp-ar to llvm-ar-mp-5.0 .

**Comment by midipix** [comment]
For now, and since you are using the doltlibtool to invoke rlibtool, change the invocation line in doltlibtool into:

`/opt/local/bin/rdlibtool --ar=llvm-ar "$@"`

That should presumably get you a notch further, or perhaps even all the way through.

**Comment by RJVB** [comment]
BTW, unrelated: what is the `config.project` file for? Is there an IDE which uses it?
(I'd write a CMake model but the project has become a little bit too complex to my taste to start doing that)

**Comment by midipix** [comment]
> BTW, unrelated: what is the `config.project` file for? Is there an IDE which uses it?
(I'd write a CMake model but the project has become a little bit too complex to my taste to start doing that)

config.project contains project-specific defaults that are used by configure. It's an easy way for a project's developer to set some settings that are unlikely to vary between build locations, or are otherwise required by the package (for instance, -D_XOPEN_SOURCE). Not sure what you mean by too complext (sofort, slibtool). I'd argue that both maintain a good balance between between feature complete and simple to read/modify. That being said, concrete suggestions are always welcome.

**Comment by midipix** [comment]
> BTW, unrelated: what is the `config.project` file for? Is there an IDE which uses it?
(I'd write a CMake model but the project has become a little bit too complex to my taste to start doing that)

Can you explain what you mean by that? From my testing, and also from your previous report regarding libcompat.la, I thought that the last commit fixed the issue with static vs. dynamic flavor when an explicit -static is present.

Otherwise, and given that you are using one tweak of another for llvm-ar (a separate issue, noted), does the build now succeed?

**Comment by RJVB** [comment]
> Not sure what you mean by too complext (sofort, slibtool).

The source tree looks too complex to sit down and write a CMake build system in the proverbial 5 minutes (CMake is great but offers little to no help in terms of automation or import features; you have to write everything by hand).

> From my testing, and also from your previous report regarding libcompat.la,
> I thought that the last commit fixed the issue with static vs. dynamic
> flavor when an explicit -static is present.

Did you actually try to build libcompat.a? As mentioned that still succeeds for me only when I remove the -rpath argument. In short, your commit is halfway there.

I'm trying to figure out what's missing.

> Otherwise, and given that you are using one tweak of another for llvm-ar (a
> separate issue, noted), does the build now succeed?

No, the build does not succeed, and my tweak is just a few lines of additional code that will obtain the ar and/or ranlib commands from environment variables if those are set. There are no side-effects when the vars are not set.

**Comment by RJVB** [comment]
Gotcha, I think: you must have forgotten about 

slbt_exec_link.c:1179
```
	/* -disable-static? */
	if (dctx->cctx->drvflags & SLBT_DRIVER_DISABLE_STATIC)
		if (dctx->cctx->rpath)
			return slbt_exec_link_create_noop_symlink(
				dctx,ectx,arfilename);
```

**Comment by midipix** [comment]
I tested the patch as such, not in the context of vlc (the build tree you sent was looking in terms of looking at the various files, but I do not have any of the prerequisites), and it worked. The check above will probably need to be modified in light of the last commit.

**Comment by midipix** [comment]
> The source tree looks too complex to sit down and write a CMake build system in the proverbial 5 minutes (CMake is great but offers little to no help in terms of automation or import features; you have to write everything by hand).

Allow me to respectfully disagree on the first part (src/*/*.[ch] is quite common, and in the absence of a concrete suggestion I'm not sure what you'd consider a simpler alternative). The second part (whether or not use cmake in a project) is a matter of preference.

In any case, and for the sake of efficiency and etiquette alike, let us focus on the original problem presented in this issue.

**Comment by midipix** [comment]
> 	/* -disable-static? */
	if (dctx->cctx->drvflags & SLBT_DRIVER_DISABLE_STATIC)

The fix should take place in the driver, will push sometime today.

**Comment by RJVB** [comment]
Anyway this wasn't meant as a critique (or maybe just on my lazyness to sit down and figure things out). And FWIW, a CMake build system would only give me a way to import the project properly into an IDE, without having to define header search paths manually etc.

Back to the business at hand, I tested with 
```
@@ -1158,7 +1177,7 @@ static int slbt_exec_link_create_archive(
 	char		arlink [PATH_MAX];
 
 	/* -disable-static? */
-	if (dctx->cctx->drvflags & SLBT_DRIVER_DISABLE_STATIC)
+	if (!(dctx->cctx->drvflags & SLBT_DRIVER_DISABLE_STATIC) && !(dctx->cctx->drvflags & SLBT_DRIVER_DISABLE_SHARED))
 		if (dctx->cctx->rpath)
 			return slbt_exec_link_create_noop_symlink(
 				dctx,ectx,arfilename);
```

and that fixed the issue. I can see how this is probably not the proper fix, the condition looks a bit weird (but apparently you do set both disable flags when -rpath and -static are used together?!).

Note also that I made another change "long ago" which still applies: when using llvm-ar I make the command use `altv` instead of `argv`. In this particular case the vectors are identical, but I labelled the patchfile that makes this change as "llvm-ar_no_mri". I have no recollection what I meant with "mri" ... :(

I should put my fork (back) on github so you can go over my various tweaks.

**Comment by midipix** [comment]
> and that fixed the issue. I can see how this is probably not the proper fix, the condition looks a bit weird (but apparently you do set both disable flags when -rpath and -static are used together?!).

Perfect; the weirdness is indeed why I said that this needs to be fixed in the driver. Good to know that this was the last(?) obstacle as far as the current build of vlc goes.

> Note also that I made another change "long ago" which still applies: when using llvm-ar I make the command use altv instead of argv. In this particular case the vectors are identical, but I labelled the patchfile that makes this change as "llvm-ar_no_mri". I have no recollection what I meant with "mri" ... :(

Yep, that one is related to the other open issue concerning llvm-ar (but not only, bsd has similar issues too). MRI scripts are supported by gnu-ar, and allow, for instance, to easily merge two archives. Merging archives internally and accordingly getting rid of that dependency altogether is still on my list of things to do...

Will update later once I've pushed the driver-specific fix.

**Comment by midipix** [comment]
As of commit 89f5ab2, conflicting and/or incompatible command-line arguments (-static, -shared, --disable-static, --disable-shared) are now properly handled.

SLBT_DRIVER_ALL_STATIC would have been the wrong solution here, as its main purpose is to allow the creation of statically-linked executable binaries.

Many thanks for the catch, and should anything else remain open (other than llvm-ar, which already has an issue) please let me know:=)

**Comment by RJVB** [comment]
On Thursday October 17 2019 11:10:07 midipix wrote:
>As of commit 89f5ab2, conflicting and/or incompatible command-line arguments (-static, -shared, --disable-static, --disable-shared) are now properly handled.

Curiously, building with the "shunt" to rlibtool on top of the libtool script *slows* down the build, even if [r]libtool is used only for linking (which however does take a significant time since I'm using LTO:

Building VLC with the full libtool script (make -kj4):
802.466 user_cpu 116.238 kernel_cpu 4:51.89 total_time 314.7%CPU {350944M 16954F 11740297R 9823080I 404335O 133236w 699286c}

Idem, with libtool handing off to rlibtool immediately:
909.319 user_cpu 67.201 kernel_cpu 5:41.02 total_time 286.3%CPU {350972M 47490F 6262674R 15618824I 1004836O 74975w 585461c}

That's almost as if there's a significant difference in the linker arguments...

**Comment by midipix** [comment]
Is it possible that gnu libtool "swallows" some of the optimization arguments, resulting in the linker doing less work when invoked by libtool script?

**Comment by RJVB** [comment]
I'd hope not, but it's something I'd have to check. I don't suppose rlibtool replaces `-flto=thin` with `-flto`? ;)

**Comment by midipix** [comment]
slibtool never performs that kind of argument substitution, but with  gnu libtool there are all sorts of transformations in place. A common one is -no-undefined, which often turns into a no-op. Perhaps you could perform full builds from scratch with both gnu libtool and rlibtool, using the following two invocation commands:

`strace -f -o libtool.log -e execve make`
`strace -f -o rlibtool.log -e execve make LIBTOOL=rlibtool MAKE="make LIBTOOL=rlibtool"`

**Comment by RJVB** [comment]
Perhaps I could indeed :) Not a very high priority for me, not until I start noticing similar things with projects that make much more intensive use of libtool. As it is you wouldn't expect a large benefit with VLC anyway given it bypasses libtol fo all compilations.

**Comment by midipix** [comment]
Closing (original problem resolved as of commit 89f5ab2). Discussion of llvm-ar support continues as part of issue #13.
